[MRG+1] check length of feature_names in export_graphviz (#8477)

aikinogard opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#Contributing-Pull-Requests
-->
#### Reference Issue
<!-- Example: Fixes #1234 -->
Fixes #8477 

#### What does this implement/fix? Explain your changes.

check length of feature_names in export_graphviz

- <s>raise ValueError if len(feature_names) > tree.n_features</s>
- <s>add unit test for len(feature_names) > tree.n_features</s>
- <s>change the comment of existing unit test for len(feature_names) < tree.n_features</s>
- raise error when length features_names != n_features
- add unit test

#### Any other comments?

<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

aikinogard committed:
#8477

check length of feature_names in export_graphviz

- raise ValueError if len(feature_names) > tree.n_features
- add unit test for len(feature_names) > tree.n_features
- change the comment of existing unit test for len(feature_names) <
tree.n_features

aikinogard commented:
The fail of check on AppVeyor
Environment: PYTHON=C:\Python35, PYTHON_VERSION=3.5.0, PYTHON_ARCH=32
FAIL: sklearn.decomposition.tests.test_pca.test_singular_values

but this is not the change I made. My changing is in sklearn.tree. This is weird.

amueller commented on sklearn/tree/tests/test_export.py:
```diff
@@ -212,10 +212,15 @@ def test_graphviz_errors():
     assert_raises(IndexError, export_graphviz, clf, out, feature_names=[])
 
+    # Check feature_names more than number of features error
+    out = StringIO()
```
Maybe use "outfile="none" instead of using StringIO.

amueller commented on sklearn/tree/tests/test_export.py:
```diff
@@ -212,10 +212,15 @@ def test_graphviz_errors():
 
+    # Check feature_names more than number of features error
+    out = StringIO()
+    assert_raises(ValueError, export_graphviz, clf, out,
```
Please check the error message with ``assert_raise_message``

amueller commented:
i think you should show a warning for either too few or too many features so that instead of an index error there is a more helpful error message.

aikinogard commented:
@amueller It is okay to show a warning for too many feature_names. However, for too few feature_names, some features will not be able to get a name. Are you suggesting to padding the feature_names list if user's input shorter than number of features in tree?

amueller commented:
No, i'm suggesting to provide a good error message whenever the length doesn't match.

aikinogard committed:
fix error and warning

- include length of feature_names and number of features in tree in the
error and warning message.
- raise error for too few feature_names
- for too much feature_names, will use the first n_features. raise an
warning for users
- use assert_raise_message and assert_warns_message in test to check
message.

aikinogard commented:
@amueller I have updated the code

- include length of feature_names and number of features in tree in the
error and warning message.
- raise error for too few feature_names
- for too much feature_names, will use the first n_features. raise an
warning for users
- use assert_raise_message and assert_warns_message in test to check
message.

aikinogard commented:
@amueller please let me know if you want any improvement for this PR. Thanks.

MechCoder commented on sklearn/tree/export.py:
```diff
@@ -216,6 +217,19 @@ def node_to_str(tree, node_id, criterion):
             if feature_names is not None:
+                # Check length of feature_names
+                # raise error for too few feature_names
+                if len(feature_names) < tree.n_features:
```
Failing early is always better.

In other words, you should check this once at the beginning when `export_graphviz` is called rather than when `node_to_str` is called. You can do that by accessing the attribute `n_features_` of the provided `decision_tree`.

aikinogard committed:
move the error and warning from node_to_str to export_graphviz so it will fail early for wrong length of feature_names

aikinogard commented:
@MechCoder Thanks for the code review. I agree with the point to make the failing happen earlier. I have made the change of moving the feature_name length check into the export_graphviz. It passed the unit test I wrote before suggested by @amueller . Please let me know if there are any other suggestions.

MechCoder commented on sklearn/tree/export.py:
```diff
@@ -397,6 +398,23 @@ def recurse(tree, node_id, criterion, parent=None, depth=0):
+                                    decision_tree.n_features_))
+            # for too much feature_names, will use the first n_features
+            # raise an warning for users
+            if len(feature_names) > decision_tree.n_features_:
```
Not sure that this is appropriate, I would just raise an error they are not equal. It is dangerous to guess that just selecting the first n values of feature_names is what the user intended.

aikinogard replied:
Before I worked on this issue, it used the first n_features in features_names by default if feature names was longer than n_features. I make it a warning to keep it consistent with the previous code.

Do you want me to raise error for length  features_names != n_features?

MechCoder replied:
Good point. My vote would be still to be raise an error and document it as a bug-fix over here (https://github.com/scikit-learn/scikit-learn/blob/master/doc/whats_new.rst)

Something on the lines of `export_graphviz` now raises an error if length of feature_names is not the same as the number of features.

aikinogard replied:
Ok. I will make the change to raise error for feature_names != n_features and modify the unit test. Thanks.

aikinogard committed:
raise error if length of feature_names does not match number of features in the decision tree

aikinogard commented:
@MechCoder Please review the new change. I raise error when length features_names != n_features and modified the unit test.

MechCoder reviewed (approved):
LGTM

MechCoder commented:
Ah did not see that Travis is faliing due to a pep8 error. Could you please look at the build and fix it?

aikinogard committed:
fix pep8

aikinogard committed:
remove unused assert_warns_message import in test_export.py

aikinogard commented:
@MechCoder I removed assert_warns_message in the import in the unit test. It should work now. For adding the bug fix doc in 
https://github.com/scikit-learn/scikit-learn/blob/master/doc/whats_new.rst
shall I add it together n this pull request or in a separate way?

MechCoder commented:
please add it over here.

aikinogard committed:
add bug fix in doc/whats_new.rst

aikinogard commented:
@MechCoder I have added the bug fix in the doc. Thank you for the review.

aikinogard commented:
I didn't see the point of conflict since both line
from sklearn.utils.testing import assert_raise_message
from sklearn.exceptions import NotFittedError
can exist.

aikinogard commented:
@amueller @MechCoder The change and unit test are finished. Please take a look on the pull request and merge it. The PR has been here for almost 2 months.

MechCoder commented:
Can you fix the merge conflicts in `sklearn/tree/tests/test_export.py`?

For 2nd review random pings: @lesteve @raghavrv (It is a small PR)

aikinogard commented:
@MechCoder It requires write access to resolve the conflict.
from sklearn.utils.testing import assert_raise_message
from sklearn.exceptions import NotFittedError
we can keep both of them in the test_export.py

MechCoder commented:
Look at the portions starting with Note: over here (http://scikit-learn.org/dev/developers/contributing.html#how-to-contribute) to know how to resolve merge conflicts.

Let me know if you need any help.

aikinogard committed:
Merge branch 'master' into issue8477

aikinogard commented:
@MechCoder I found I can solve the conflict directly on github. :-)
The conflict is resolved. Thanks!
@lesteve @raghavrv please review it at your convenience.

aikinogard commented:
@jnothman @TomDLT could you review my PR at your convenience?

raghavrv commented on doc/whats_new.rst:
```diff
@@ -276,6 +276,11 @@ Bug fixes
      integer classes. :issue:`8676` by :user:`Vathsala Achar <VathsalaAchar>`.
 
+   - Fix a bug where :func:`sklearn.tree.export_graphviz` will raise an error
+     when the length features_names does not match n_features in the decision
```
length of

raghavrv replied:
nitpick: backticks for feature_names and n_features

raghavrv commented on doc/whats_new.rst:
```diff
@@ -276,6 +276,11 @@ Bug fixes
    - Fixed a bug in :class:`svm.OneClassSVM` where it returned floats instead of
      integer classes. :issue:`8676` by :user:`Vathsala Achar <VathsalaAchar>`.
 
+   - Fix a bug where :func:`sklearn.tree.export_graphviz` will raise an error
```
nitpick again: past tense is good I think - ...raised an error...

raghavrv commented on doc/whats_new.rst:
```diff
@@ -276,6 +276,11 @@ Bug fixes
+   - Fix a bug where :func:`sklearn.tree.export_graphviz` will raise an error
+     when the length features_names does not match n_features in the decision
+     tree.
+     :issue:`8512` by `Li Li <https://github.com/aikinogard>`_.
```
Can you use this syntax? https://github.com/scikit-learn/scikit-learn/pull/8512/files#diff-3433b590ca8611f80d7687b51a869383R277

raghavrv commented on sklearn/tree/export.py:
```diff
@@ -399,6 +402,16 @@ def recurse(tree, node_id, criterion, parent=None, depth=0):
             return_string = True
             out_file = six.StringIO()
 
+        # Check length of feature_names before get into the tree node
```
nickpick (feel free to ignore): before getting into the...

raghavrv commented on sklearn/tree/export.py:
```diff
@@ -399,6 +402,16 @@ def recurse(tree, node_id, criterion, parent=None, depth=0):
+        # n_features_ in the decision_tree
+        if feature_names is not None:
+            if len(feature_names) != decision_tree.n_features_:
+                raise ValueError("Length of feature_names=%d "
```
Length of feature_name, %d

raghavrv commented on sklearn/tree/export.py:
```diff
@@ -330,7 +332,8 @@ def recurse(tree, node_id, criterion, parent=None, depth=0):
                         colors['bounds'] = (np.min(-tree.impurity),
                                             np.max(-tree.impurity))
-                    elif tree.n_classes[0] == 1 and len(np.unique(tree.value)) != 1:
+                    elif tree.n_classes[0] == 1 and \
```
see if you can use a bracket and avoid `\`

raghavrv commented on sklearn/tree/tests/test_export.py:
```diff
@@ -218,9 +219,19 @@ def test_graphviz_errors():
-    # Check feature_names error
-    out = StringIO()
-    assert_raises(IndexError, export_graphviz, clf, out, feature_names=[])
+    # Check feature_names less than number of features error
```
--> Check if it errors when length of feature_names mismatches with number of features

raghavrv commented on sklearn/tree/tests/test_export.py:
```diff
@@ -218,9 +219,19 @@ def test_graphviz_errors():
+    out = None
+    message = "Length of feature_names=3 does not " + \
+              "match number of features=2"
+    assert_raise_message(ValueError, message, export_graphviz, clf, out,
```
directly pass `out=None`

raghavrv commented on sklearn/tree/tests/test_export.py:
```diff
@@ -218,9 +219,19 @@ def test_graphviz_errors():
+    out = None
+    message = "Length of feature_names=1 does not " + \
+              "match number of features=2"
+    assert_raise_message(ValueError, message, export_graphviz, clf, out,
```
directly pass `out=None`

raghavrv commented on sklearn/tree/tests/test_export.py:
```diff
@@ -218,9 +219,19 @@ def test_graphviz_errors():
-    assert_raises(IndexError, export_graphviz, clf, out, feature_names=[])
+    # Check feature_names less than number of features error
+    out = None
+    message = "Length of feature_names=1 does not " + \
```
use bracket and avoid `\`

raghavrv commented on sklearn/tree/tests/test_export.py:
```diff
@@ -218,9 +219,19 @@ def test_graphviz_errors():
+
+    # Check feature_names greater than number of features warning
+    out = None
+    message = "Length of feature_names=3 does not " + \
```
This needs to be fixed after the changes above...

raghavrv reviewed (changes requested):
Thanks @aikinogard... LGTM pending (mostly) nitpicks...

aikinogard committed:
fix the english in doc/whats_new.rst

aikinogard committed:
fix the format and english in sklearn/tree/export.py

aikinogard committed:
fix contributor format in doc/whats_new.rst

aikinogard committed:
fix english, use bracket and avoid \ in error message

aikinogard committed:
fix conflict

aikinogard committed:
fix pep8

raghavrv commented:
Thanks, will merge once CIs turn green.

aikinogard committed:
fix pep8

aikinogard commented:
@raghavrv CIs turned green. Please merge it at your convenience. Thanks!

raghavrv commented:
:tada:
